src/control/server/server.go

```diff
@@ -207,270 +101,235 @@ func Start(log *logging.LeveledLogger, cfg *config.Server) error {
-					log.Errorf("failed to mark rank %d as dead: %s", evt.Rank, err)
-					return
-				}
-				mgmtSvc.reqGroupUpdate(ctx)
-			}
-		}))
+	srv.sysdb.ConfigureTransport(srv.grpcServer, tSec)
 
-		return nil
-	})
-	sysdb.OnLeadershipLost(func() error {
-		log.Infof("MS leader no longer running on %s", hostname())
+	return nil
+}
 
-		// Stop handling received forwarded (in addition to local)
-		// events and start forwarding events to the new MS leader.
-		eventPubSub.Reset()
-		eventPubSub.Subscribe(events.RASTypeAny, eventLogger)
-		eventPubSub.Subscribe(events.RASTypeStateChange, eventForwarder)
+func (srv *server) registerEvents() {
+	registerInitialSubscriptions(srv)
 
+	srv.sysdb.OnLeadershipGained(func(ctx context.Context) error {
+		srv.log.Infof("MS leader running on %s", hostname())
+		srv.mgmtSvc.startJoinLoop(ctx)
+		registerLeaderSubscriptions(srv)
+		return nil
+	})
+	srv.sysdb.OnLeadershipLost(func() error {
+		srv.log.Infof("MS leader no longer running on %s", hostname())
+		registerFollowerSubscriptions(srv)
 		return nil
 	})
+}
 
+func (srv *server) start(ctx context.Context, shutdown context.CancelFunc) error {
 	go func() {
-		_ = grpcServer.Serve(lis)
+		_ = srv.grpcServer.Serve(srv.listener)
 	}()
-	defer grpcServer.Stop()
+	defer srv.grpcServer.Stop()
 
-	log.Infof("%s v%s (pid %d) listening on %s", build.ControlPlaneName, build.DaosVersion, os.Getpid(), controlAddr)
+	srv.log.Infof("%s v%s (pid %d) listening on %s", build.ControlPlaneName,
+		build.DaosVersion, os.Getpid(), srv.ctlAddr)
 
 	sigChan := make(chan os.Signal)
 	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGTERM)
 	go func() {
 		// SIGKILL I/O Engine immediately on exit.
 		// TODO: Re-enable attempted graceful shutdown of I/O Engines.
 		sig := <-sigChan
-		log.Debugf("Caught signal: %s", sig)
+		srv.log.Debugf("Caught signal: %s", sig)
 
 		shutdown()
 	}()
 
-	return errors.Wrapf(harness.Start(ctx, sysdb, eventPubSub, cfg), "%s exited with error", build.DataPlaneName)
+	return errors.Wrapf(srv.harness.Start(ctx, srv.sysdb, srv.pubSub, srv.cfg),
+		"%s exited with error", build.DataPlaneName)
+}
+
+// Start is the entry point for a daos_server instance.
+func Start(log *logging.LeveledLogger, cfg *config.Server) error {
+	fd, err := processConfig(log, cfg)
+	if err != nil {
+		return err
+	}
+
+	// Create the root context here. All contexts should inherit from this one so
+	// that they can be shut down from one place.
+	ctx, shutdown := context.WithCancel(context.Background())
+	defer shutdown()
+
+	srv, err := newServer(ctx, log, cfg, fd)
+	if err != nil {
+		return err
+	}
+	defer srv.shutdown()
+
+	ctxNet, shutdownNet, err := srv.initNetwork(ctx)
+	if err != nil {
+		return err
+	}
+	defer shutdownNet()
+
+	if err := srv.initStorage(); err != nil {
+		return err
+	}
+
+	if err := srv.addEngines(ctxNet); err != nil {
+		return err
+	}
+
+	if err := srv.discoverStorage(); err != nil {
```

Comment:
I guess there isn't a clean way to run this before addEngines, right? Just thinking that there's a lot of fairly heavyweight setup in addEngines that we have to go through before we could still fail in discoverStorage.